---
title: "Quality Control Analysis"
output: html_notebook
---
```{r}
library(Matrix)
library(dplyr)
library(Seurat)
library(ggplot2)
```

```{r}
# Import file
pbmc2 <- readRDS("/Users/laurenstein/Desktop/local_data/scRNAseq_10x_PBMC/data/pbmc_filtered_seurat.rds")
pbmc2
```

```{r}
metadata <- pbmc2@meta.data
```

Expected number of cells vs. actual number of cells
Expected: 2700 
Actual: 2700

```{r}
# Rename columns
metadata <- metadata %>%
        dplyr::rename(seq_folder = orig.ident,
                      nUMI = nCount_RNA,
                      nGene = nFeature_RNA,
                      mitoRatio = mitoRatio)
```

```{r}
# Visualize the number UMIs/transcripts per cell
#Indicates how deeply sequenced the samples were
metadata %>% 
  	ggplot(aes(x=nUMI)) + 
    geom_histogram(aes(y=..density..), color = "black", fill = "white") +
  	geom_density(alpha = 0.2, fill = "purple") + 
  	scale_x_log10() + 
  	theme_classic() +
  	ylab("Cell density") +
  	geom_vline(xintercept = 500)

ggsave("/Users/laurenstein/Desktop/code-a-rama/scRNAseq_10x_PBMC/plots/QC/UMIperCell.png")
```

```{r}
# Visualize the distribution of genes detected per cell via histogram
metadata %>% 
  	ggplot(aes(x=nGene)) + 
  	geom_histogram(aes(y=..density..), color = "black", fill = "white") +
  	geom_density(alpha = 0.2, fill = "purple") + 
  	theme_classic() +
  	scale_x_log10() + 
  	geom_vline(xintercept = 300)

ggsave("/Users/laurenstein/Desktop/code-a-rama/scRNAseq_10x_PBMC/plots/QC/GenesperCell.png")
```

```{r}
# Visualize the distribution of genes detected per cell via boxplot
metadata %>% 
  	ggplot(aes(y=log10(nGene))) + 
  	geom_boxplot() + 
  	theme_classic() +
  	theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  	theme(plot.title = element_text(hjust=0.5, face="bold")) +
  	ggtitle("NCells vs NGenes")

ggsave("/Users/laurenstein/Desktop/code-a-rama/scRNAseq_10x_PBMC/plots/QC/GenesperCell_Boxplot.png")
```

```{r}
# Visualize the distribution of mitochondrial gene expression detected per cell
metadata %>% 
  	ggplot(aes(x=percent.mt)) + 
  	geom_histogram(aes(y=..density..), color = "black", fill = "white") +
  	geom_density(alpha = 0.2, fill = "purple") + 
  	scale_x_log10() + 
  	theme_classic() +
  	geom_vline(xintercept = 0.2)

ggsave("/Users/laurenstein/Desktop/code-a-rama/scRNAseq_10x_PBMC/plots/QC/MitoperCell.png")
```

```{r}
# Visualize the overall complexity of the gene expression by visualizing the genes detected per UMI 
# Look into novelty score, reference states it should be about 0.8
metadata %>%
  	ggplot(aes(x=log10GenesPerUMI)) +
  	geom_density(alpha = 0.2) +
  	theme_classic() +
  	geom_vline(xintercept = 0.8)

ggsave("/Users/laurenstein/Desktop/code-a-rama/scRNAseq_10x_PBMC/plots/QC/GenesperUMI.png")
```
MULTIVARIATE QC ANALYSIS
```{r}
# Visualize the correlation between genes detected and number of UMIs and determine whether strong presence of cells with low numbers of genes/UMIs
metadata %>% 
  	ggplot(aes(x=nUMI, y=nGene, color=mitoRatio)) + 
  	geom_point() + 
	  scale_colour_gradient(low = "purple", high = "yellow") +
  	stat_smooth(method=lm) +
  	scale_x_log10() + 
  	scale_y_log10() + 
  	theme_classic() +
  	geom_vline(xintercept = 500) +
  	geom_hline(yintercept = 250)
ggsave("/Users/laurenstein/Desktop/code-a-rama/scRNAseq_10x_PBMC/plots/QC/GenesCorrUMI.png")

```
```{r}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("DropletUtils")
```
```{r}
library(DropletUtils)
```

```{r}
set.seed(1000)
mol.info.file <- DropletUtils:::simBasicMolInfo(tempfile())
mol.info.file
```
```{r}
## -----------------------------------------------------------------------------
mol.info <- read10xMolInfo(mol.info.file)
mol.info
```

```{r}
## -----------------------------------------------------------------------------
set.seed(100)
no.sampling <- downsampleReads(mol.info.file, prop=1)
sum(no.sampling)
with.sampling <- downsampleReads(mol.info.file, prop=0.5)
sum(with.sampling)

```

```{r}
## -----------------------------------------------------------------------------
set.seed(0)
my.counts <- DropletUtils:::simCounts()
```

```{r}
## -----------------------------------------------------------------------------
br.out <- barcodeRanks(my.counts)

# Making a plot.
plot(br.out$rank, br.out$total, log="xy", xlab="Rank", ylab="Total")
o <- order(br.out$rank)
lines(br.out$rank[o], br.out$fitted[o], col="red")

abline(h=metadata(br.out)$knee, col="dodgerblue", lty=2)
abline(h=metadata(br.out)$inflection, col="forestgreen", lty=2)
legend("bottomleft", lty=2, col=c("dodgerblue", "forestgreen"), 
    legend=c("knee", "inflection"))


```
```{r}
## -----------------------------------------------------------------------------
set.seed(100)
e.out <- emptyDrops(my.counts)
e.out
```
```{r}
## -----------------------------------------------------------------------------
is.cell <- e.out$FDR <= 0.01
sum(is.cell, na.rm=TRUE)
```
```{r}
## -----------------------------------------------------------------------------
table(Limited=e.out$Limited, Significant=is.cell)

## -----------------------------------------------------------------------------
plot(e.out$Total, -e.out$LogProb, col=ifelse(is.cell, "red", "black"),
    xlab="Total UMI count", ylab="-Log Probability")


```
```{r}
set.seed(10000)

# Simulating empty droplets:
nbarcodes <- 1000
nhto <- 10
y <- matrix(rpois(nbarcodes*nhto, 20), nrow=nhto)

# Simulating cells:
ncells <- 100
true.sample <- sample(nhto, ncells, replace=TRUE)
y[cbind(true.sample, seq_len(ncells))] <- 1000

# Simulating doublets:
ndoub <- ncells/10
next.sample <- (true.sample[1:ndoub]  + 1) %% nrow(y)
next.sample[next.sample==0] <- nrow(y)
y[cbind(next.sample, seq_len(ndoub))] <- 500
```

```{r}
hto.calls <- emptyDrops(y, lower=500)
has.cell <- hto.calls$FDR <= 0.001
summary(has.cell)
```

```{r}
demux <- hashedDrops(y[,which(has.cell)], 
    ambient=metadata(hto.calls)$ambient)
demux

```

```{r}
table(demux$Best[demux$Confident])
```

```{r}
## -----------------------------------------------------------------------------
colors <- ifelse(demux$Confident, "black",
    ifelse(demux$Doublet, "red", "grey"))
plot(demux$LogFC, demux$LogFC2, col=colors,
    xlab="Log-fold change between best and second HTO",
    ylab="Log-fold change between second HTO and ambient")

## -----------------------------------------------------------------------------
set.seed(1000)
mult.mol.info <- DropletUtils:::simSwappedMolInfo(tempfile(), nsamples=3)
mult.mol.info

## -----------------------------------------------------------------------------
s.out <- swappedDrops(mult.mol.info, min.frac=0.9)
length(s.out$cleaned)
class(s.out$cleaned[[1]])

## -----------------------------------------------------------------------------
out <- chimericDrops(mult.mol.info[1])
class(out)
```

